File: advancedDataStructureImplementation/disjointSet.cpp
#include <bits/stdc++.h>
using namespace std;
class disjointSet{
    unordered_map<int,int> ump;
    public:
    disjointSet(){}
    disjointSet(vector<int>& arr){
        for(auto num:arr) ump[num]=-1;
    }
    int findSet(int u){
        int r=u;
        while(ump[r]!=-1){
            r=ump[r];
        }
        while(u!=r){
            int par=ump[u];
            ump[u]=r;
            u=par;
        }
        return r;
    }
    void setUnion(int u,int v){
        int uroot=findSet(u);
        int vroot=findSet(v);
        if(uroot == vroot) return ;
        int uchild=ump[uroot];
        int vchild=ump[vroot];
        int totalChild=uchild+vchild;
        
        if(uchild>vchild){
            ump[uroot]=vroot;
            ump[vroot]=totalChild;
        }else{
            ump[vroot]=uroot;
            ump[uroot]=totalChild;
        }
        return;
    }
    void print(){
        for(auto &pr:ump){
            cout<<"{"<<pr.first<<"<="<<pr.second<<"}"<<endl;
        }
    }
};
int main() {
	// your code goes here
	vector<int> arr;
	arr.push_back(1);
	arr.push_back(2);
	arr.push_back(3);
	arr.push_back(4);
	disjointSet ds=*new disjointSet(arr);
	
	ds.print();
	return 0;
}

File: modules/java-modules/com/library/heavyds/DisjointSet.java
package com.library.heavyds;

public class DisjointSet{
    public int [] arr;
    public int n;
    public DisjointSet(int n){
        this.n=n;
        arr=new int[n];
        for(int i=0;i<n;i++) arr[i]=-1;
    }
    public int findSet(int u){
        int r=u;
        while(arr[r] >= 0){
            r=arr[r];
        }
        while(u != r){
            int next=arr[u];
            arr[u]=r;
            u=next;
        }
        return r;
    }
    public boolean setUnion(int u,int v){
        int uroot=this.findSet(u);
        int vroot=this.findSet(v);
        if(uroot==vroot) return false;
        int totalChilds=arr[uroot]+arr[vroot];
        if(arr[uroot]<arr[vroot]){
            arr[vroot]=uroot;
            arr[uroot]=totalChilds;
        }else{
            arr[uroot]=vroot;
            arr[vroot]=totalChilds;
        }
        return true;
    }
    public boolean preferedUnion(int u,int v){
        int uroot=this.findSet(u);
        int vroot=this.findSet(v);
        if(uroot==vroot) return false;
        int totalChilds=arr[uroot]+arr[vroot];
        arr[uroot]=vroot;
        arr[vroot]=totalChilds;
        return true;
    }
}

